### PR TITLE
fix datasource not accurate

### PR DESCRIPTION
**Pull request by wonderflow**
## Changes
   
  - [ ] feature1
  - [ ] feature2 
  - [ ] fixbug1
  - [ ] fixbug2
 
## Reviewers

  - [ ] @[someone] please review
  - [ ] @[someotherone] please review

## Wiki Changes
 
  - options1...
  - options2...
    
## Checklist
   
   - [ ] Rebased/mergable
   - [ ] Tests pass
   - [ ] Wiki updated

### CODE REVIEW

File: mgr/mgr.go [hunk 1]
```diff
@@ -716,14 +716,18 @@ func (m *Manager) StartRunner(name string) (err error) {
...
+			m.setRunnerConfig(filename, conf)
 		}
 	}
 	return
 }
 
+func (m *Manager) setRunnerConfig(filename string, conf RunnerConfig) {
+	m.lock.Lock()
+	m.runnerConfig[filename] = conf
+	m.lock.Unlock()
```
Comment (redHJ) [inline mgr/mgr.go]: 这边不用defer m.lock.unlock() ? readRunners里用的是defer

File: reader/bufreader.go [hunk 2]
```diff
@@ -67,6 +67,11 @@ type BufReader struct {
 	statsLock sync.RWMutex
 
 	lastErrShowTime time.Time
+
+	// 这里的变量用于记录buffer中的数据从底层的那个DataSource出来的，用于精准定位seqfile的DataSource
```
Comment (redHJ) [inline reader/bufreader.go]: 哪个

File: reader/seqfile.go [hunk 3]
```diff
@@ -307,6 +322,7 @@ func (sf *SeqFile) Read(p []byte) (n int, err error) {
 						sf.newLineNotAdded = true
 					}
 				}
+				sf.justOpenedNewFile = true
```
Comment (redHJ) [inline reader/seqfile.go]: 放在sf.open(fi)后？

File: mgr/mgr.go (hunk 1 above)
Comment (wonderflow) [reply mgr/mgr.go]: 这个一样的

File: reader/bufreader.go [hunk 4]
```diff
@@ -175,19 +180,37 @@ func (b *BufReader) fill() {
...
 		}
+		if b.lastRdSource != b.rd.Source() {
+			//这个情况表示文件的数据源出现了变化，在buf中已经出现了2个数据源的数据，要定位是哪个位置的数据出现的分隔
+			if rc, ok := b.rd.(NewLineBytesRecorder); ok {
+				idx := rc.NewLineBytesIndex()
+				// 从 NewLineBytesIndex 函数中返回的index值就是本次读取的批次中上一个DataSource的数据量，加上b.w就是上个DataSource的整体数据
+				b.lastRdByteIndex = b.w + idx
+			}
+		}
+
```
[the remark this answers is not in the document]
Comment (wonderflow) [reply reader/bufreader.go]: 必须放在 b.w+=n之前，b.lastRdSource在上面会被更新

File: reader/seqfile.go (hunk 3 above)
Comment (wonderflow) [reply reader/seqfile.go]: 可以

Comment (wonderflow) [reply reader/seqfile.go]: 可以

**Pull request merged**